### PR TITLE
Fix/watch

### LINKED ISSUE #166: [Bug]: Watch Concurrent Bug
### Description about the bug

When we call `watch` in `kvwatcher.rs`, the following things will happen:

1. get events from the requested revision
2. add watcher to the map and start listening for events

However, these two steps are not atomic operation. So if a new event occur after 1 but before 2, then we will lose this event.

### Version

0.1.0 (Default)

### Relevant log output

```shell
Not yet produced.
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by themanforfree**
Fix #166 , and fix the atomicity of index and db operations

### CODE REVIEW

File: xline/src/storage/kv_store.rs [hunk 1]
```diff
@@ -436,14 +442,15 @@ where
     }
 
     /// Get `KeyValue` start from a revision and convert to `Event`
-    pub(crate) fn get_event_from_revision(
+    pub(crate) fn get_events_from_revision(
         &self,
         key_range: KeyRange,
         revision: i64,
-    ) -> Result<Vec<Event>, ExecuteError> {
+    ) -> Result<(i64, Vec<Event>), ExecuteError> {
```
Comment (rogercloud) [inline xline/src/storage/kv_store.rs:449]: Please comment the meaning of the return value as it's complicated.

File: xline/src/server/watch_server.rs [hunk 2]
```diff
@@ -49,9 +49,9 @@ where
         ST: Stream<Item = Result<WatchRequest, tonic::Status>> + Unpin,
         W: KvWatcherOps,
     {
-        let (event_tx, event_rx) = mpsc::channel(CHANNEL_SIZE);
+        let (event_tx, mut event_rx) = mpsc::channel(CHANNEL_SIZE);
         let (stop_tx, stop_rx) = flume::bounded(0);
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/watch_server.rs]: Fixed

File: xline/src/storage/kvwatcher.rs [hunk 3]
```diff
@@ -244,33 +260,61 @@ where
...
+    /// Mark as syncing done and notify
+    fn sync_done(&self) {
+        self.syncing.store(false, Ordering::SeqCst);
+        self.syncing_notify.notify(1);
+    }
+
+    /// If current state is syncing, wait until it is done
+    async fn wait_syncing(&self) {
+        while self.is_syncing() {
+            self.syncing_notify.listen().await;
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/kvwatcher.rs]: Fixed

File: xline/src/storage/kv_store.rs (hunk 1 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:435]: Fixed

File: xline/src/storage/revision.rs [hunk 4]
```diff
@@ -11,6 +11,9 @@ pub(crate) struct KeyRevision {
     pub(crate) mod_revision: i64,
     /// Sub revision in one transaction
     pub(crate) sub_revision: i64,
+    /// Available flag
+    // true means this revision is written to storage
```
Comment (Phoenix500526) [inline xline/src/storage/revision.rs:15]: IMO, using the present perfect tense here would be more straightforward, like "/// true means this revision has been already written to storage".

File: xline/src/storage/index.rs [hunk 5]
```diff
@@ -33,7 +45,7 @@ impl Index {
     fn get_revision(revs: &[KeyRevision], revision: i64) -> Option<Revision> {
         // TODO: handle future revision
```
Comment (Phoenix500526) [inline xline/src/storage/index.rs:46]: Shall we handle this TODO in this pr?

File: xline/src/storage/index.rs [hunk 6]
```diff
@@ -91,36 +131,42 @@ pub(crate) trait IndexOperate {
...
         match RangeType::get_range_type(key, range_end) {
-            RangeType::OneKey => index
+            RangeType::OneKey => inner
+                .index
                 .get(key)
                 .map(|revs| Self::filter_revision(revs, revision))
                 .unwrap_or_default(),
-            RangeType::AllKeys => index
+            RangeType::AllKeys => inner
+                .index
```
Comment (Phoenix500526) [inline xline/src/storage/index.rs:164]: The test case `test_get` only covers the `RangeType::OneKey` match arm, please enrich this test case to cover all match arms.

File: xline/src/storage/index.rs [hunk 7]
```diff
@@ -49,6 +61,34 @@ impl Index {
...
+        sub_revision: i64,
+    ) -> Option<(Revision, Revision)> {
+        let last_rev = Self::get_revision(revs, 0)?;
+        let del_rev = KeyRevision::new_deletion(revision, sub_revision);
+        revs.push(del_rev);
+        Some((last_rev, del_rev.as_revision()))
+    }
+
+    /// Mark the `KeyRevision` as available
+    pub(crate) fn mark_available(&self, revision: i64) {
```
Comment (Phoenix500526) [inline xline/src/storage/index.rs:78]: The `mark_available` is only being used in the `kv_store`, so there is no point to describe it as crate visibility, super is enough.

File: xline/src/storage/index.rs [hunk 8]
```diff
@@ -134,45 +180,53 @@ impl IndexOperate for Index {
...
+                    .filter_map(|revs| Self::gen_del_revision(revs, revision, sub_revision))
+                    .collect();
+                let keys = if pairs.is_empty() {
+                    vec![]
+                } else {
+                    vec![key.to_vec()]
+                };
+                (pairs, keys)
+            }
+            RangeType::AllKeys => {
```
Comment (Phoenix500526) [inline xline/src/storage/index.rs:200]: Same as the `get_from_rev`. Plz enrich the test case `test_delete` to cover the `RangeType::AllKeys` and `RangeType::Range` match arms.

File: xline/src/server/command.rs [hunk 9]
```diff
@@ -278,6 +278,7 @@ where
             res = sync_res;
         }
         self.persistent.flush_ops(ops)?;
+        self.kv_storage.mark_index_available(res.revision());
```
Comment (Phoenix500526) [inline xline/src/server/command.rs:281]: The `mark_index_available` will panic. Shall we deal with index error in this way? WDYT? @rogercloud

File: xline/src/storage/kv_store.rs [hunk 10]
```diff
@@ -937,6 +944,33 @@ mod test {
...
+                    );
+                    let (_res, ops) = store.after_sync(&req).await.unwrap();
+                    db.flush_ops(ops).unwrap();
+                }
+            }
+        });
+        tokio::time::sleep(std::time::Duration::from_micros(500)).await;
+        let revs = store.index.get_from_rev(b"foo", b"", 1);
+        let kvs = store.get_values(&revs).unwrap();
+        assert_eq!(kvs.len(), revs.len());
```
Comment (Phoenix500526) [inline xline/src/storage/kv_store.rs:971]: Please add some concrete diagnostic messages for this assertion. Additionally, I don't get the point why this test can prove that the kv store process is atomic.

File: xline/src/storage/revision.rs (hunk 4 above)
Comment (themanforfree) [reply xline/src/storage/revision.rs]: Fixed

File: xline/src/storage/index.rs (hunk 6 above)
Comment (themanforfree) [reply xline/src/storage/index.rs:165]: Fixed

File: xline/src/storage/index.rs (hunk 7 above)
Comment (themanforfree) [reply xline/src/storage/index.rs]: Fixed

File: xline/src/storage/index.rs (hunk 8 above)
Comment (themanforfree) [reply xline/src/storage/index.rs:203]: Fixed

File: xline/src/storage/kv_store.rs (hunk 10 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs]: Fixed, If the kv store is not atomic, some kvs maybe insert to index, but not flush to db, so `revs.len()` will lager than `kvs.len()`

File: xline/src/storage/index.rs (hunk 5 above)
Comment (themanforfree) [reply xline/src/storage/index.rs]: Added

File: xline/src/storage/kv_store.rs [hunk 11]
```diff
@@ -922,8 +928,9 @@ mod test {
         );
         let db = DB::open(&StorageConfig::Memory)?;
         let store = init_store(db).await?;
-        let (_ignore, ops) = store.after_sync(&txn_req).await?;
+        let (sync_res, ops) = store.after_sync(&txn_req).await?;
         store.db.flush_ops(ops)?;
```
Comment (rogercloud) [inline xline/src/storage/kv_store.rs:932]: If these two operations always happend togethor, put them in one function to force the "pair"?

File: xline/src/storage/kv_store.rs [hunk 12]
```diff
@@ -937,6 +944,37 @@ mod test {
         Ok(())
     }
 
+    #[tokio::test(flavor = "multi_thread", worker_threads = 10)]
+    async fn test_kv_store_atomic() {
```
Comment (rogercloud) [inline xline/src/storage/kv_store.rs:948]: Atomic usually means we can't insert other operations between the continuous operations in the logic view. I don't think the test tells the store is "atomic".

File: xline/src/storage/kv_store.rs (hunk 11 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:932]: they always happened together in this module's tests, but not else where, I added a function to reduce duplication

File: xline/src/storage/kv_store.rs (hunk 12 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:948]: Renamed

File: xline/src/server/command.rs (hunk 9 above)
Comment (themanforfree) [reply xline/src/server/command.rs:281]: we can't recover it, all keys in `unavailable_cache` will be inserted into `index` first and then into `unavailable_cache`, so it is impossible to read none here

Comment (Phoenix500526) [reply xline/src/server/command.rs:281]: Then why don't you use the `unreachable!` macro to replace the `panic!`?

Comment (themanforfree) [reply xline/src/server/command.rs:281]: Fixed

File: xline/src/storage/kv_store.rs [hunk 13]
```diff
@@ -960,15 +994,23 @@ mod test {
...
             );
-            let _cmd_res = store.execute(&req)?;
-            let (_sync_res, ops) = store.after_sync(&req).await?;
-            store.db.flush_ops(ops)?;
+            sync_req(&store, &req).await?;
         }
         Ok(store)
     }
 
+    async fn sync_req(
```
Comment (rogercloud) [inline xline/src/storage/kv_store.rs:1002]: the inner operations are not "syncing" but, execute "after_sync" step and flush to persistent store. Shall we rename it "exe_as_and_flush"?

File: xline/src/storage/kvwatcher.rs [hunk 14]
```diff
@@ -258,50 +268,50 @@ where
...
-                    .flatten()
-                    .collect();
-                for watcher in watchers {
-                    #[allow(clippy::indexing_slicing)]
-                    watcher_events
-                        .entry(watcher)
-                        .or_default()
-                        .push(event.clone());
+    fn handle_kv_updates(&self, (revision, all_events): (i64, Vec<Event>)) {
+        let watcher_map_r = self.watcher_map.read();
```
Comment (rogercloud) [inline xline/src/storage/kvwatcher.rs:279]: Why do we remove the "map_read" function call?

File: xline/src/storage/kvwatcher.rs [hunk 15]
```diff
@@ -332,3 +342,84 @@ impl WatchEvent {
...
+        }
+
+        assert_eq!(map.len(), 100);
+        for (k, count) in map {
+            assert_eq!(count, 1, "key {k} should be notified once");
+        }
+        handle.abort();
+    }
+
+    fn init_empty_store() -> (Arc<KvStore<DB>>, Arc<DB>, Arc<KvWatcher<DB>>) {
```
Comment (rogercloud) [inline xline/src/storage/kvwatcher.rs:409]: Shall we move the util function to the top of test mod?

Comment (themanforfree) [reply xline/src/storage/kvwatcher.rs]: Moved

File: xline/src/storage/kv_store.rs (hunk 13 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs]: Renamed

File: xline/src/storage/index.rs [hunk 16]
```diff
@@ -49,6 +61,34 @@ impl Index {
...
+        sub_revision: i64,
+    ) -> Option<(Revision, Revision)> {
+        let last_rev = Self::get_revision(revs, 0)?;
+        let del_rev = KeyRevision::new_deletion(revision, sub_revision);
+        revs.push(del_rev);
+        Some((last_rev, del_rev.as_revision()))
+    }
+
+    /// Mark the `KeyRevision` as available
+    pub(super) fn mark_available(&self, revision: i64) {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/index.rs:80]: Commented

File: xline/src/server/kv_server.rs [hunk 17]
```diff
@@ -198,13 +198,19 @@ where
...
     /// Validate range request before handle
-    fn check_range_request(req: &RangeRequest) -> Result<(), tonic::Status> {
+    fn check_range_request(req: &RangeRequest, current_revision: i64) -> Result<(), tonic::Status> {
         if req.key.is_empty() {
             return Err(tonic::Status::invalid_argument("key is not provided"));
         }
         if !SortOrder::is_valid(req.sort_order) || !SortTarget::is_valid(req.sort_target) {
             return Err(tonic::Status::invalid_argument("invalid sort option"));
         }
+        if req.revision > current_revision {
```
Comment (Phoenix500526) [inline xline/src/server/kv_server.rs:208]: Why do we need to check the revision here now when we didn't need to before?

Comment (themanforfree) [reply xline/src/server/kv_server.rs:208]: it was a todo before
https://github.com/datenlord/Xline/pull/247#discussion_r1190788475

**Review by Phoenix500526 — APPROVED** [review]

**Review by rogercloud — APPROVED** [review]

**Pull request merged**